Added Text on Json File and Small Design Change on Summaries and Options sub dialogs on Column Summaries Dialog.

MeSophie opened the pull request:
Fixes Partially #8767
Fixes Partially #8768
@rdstern @N-thony I made added text on json file and made some small design change on Summaries and Options sub dialogs. I also correct the translation of `Maximum Number of Missing Allowed` in Crowdin. Please have a look.

MeSophie commented:
@rdstern Please anything new about this PR?

rdstern reviewed (commented):
@MeSophie that's all good so far.  The JSON title confused me, because I usually don't check those ones.
I am very happy that the layout is now fixed on the hydroGOF.
The extra which max and which min are also working fine.
The remaining bit to add is to be able to display another variable instead of the row value.

I show the corresponding improvement in the summary dialog:
![image](https://github.com/IDEMSInternational/R-Instat/assets/11226469/99a01406-f481-4ad7-a1c9-c04c5f4b8755)

You have added the which_max and which_min, but the other variable is not enabled for those options.  It should be, and then it gives the where_max and where_min.  Currently it is enabled when another option is also chosen and that seems not to work so well now.

But I am very happy to see those improvements already.

MeSophie commented:
@rdstern I made some change according to your comment. I implement the Where functions according to the description I saw on calculator buttons. Please have a look.

rdstern commented:
@MeSophie that's in the right direction, but it isn't picking up the right variable for the `where max, or min`

Here, from the calculator, is what I should get, (for the trivial case).  The where max gives me the village with the lowest (or highest) yields.  

![image](https://github.com/IDEMSInternational/R-Instat/assets/11226469/8986a3b7-354a-4437-ab2b-17fa85e73d95)

Here I tried the same, in your dialog, for which variety gave the extreme yields - but it didn't show the variety variable.

![image](https://github.com/IDEMSInternational/R-Instat/assets/11226469/5f184ae9-000c-4684-b8ca-6213684dc876)

In the results, I notice you have the `order by variety` in the command, but the `where `displays the yields, rather than the varieties.

I also note that the which_max now shows the row number within the village factor.  I think before it showed the overall row number?

MeSophie commented:
@rdstern the which_max was showed the row number within the village since I tried to modify the code to get the result for the overall row number without achieving anything. Maybe @lilyclements can help here this is the actual code:
```
summary_which_max <- function (x, na.rm = TRUE, na_type = "", ...) {
  #TODO This prevents warning and -Inf being retured. Is this desirable?
  if(length(x)==0 || (na.rm && length(x[!is.na(x)])==0)) return(NA)
  if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
  else{
    return(which.max(x))
  } 
}
```
And I wonder why we need a where.max function since we have the variable `summarize by` in the result of which.max. Unless the aim here is to obtain a unique value with the which.max function, as is the case with the calculator. If this is not the case, the columns will be repeated. Is this really necessary? I think the trick here is to get the row index for the overall row number.

Concerning the `ordered by ` variable, I noticed that it is a parameter of  `first`, `last`, and `nth`  functions which is not the case for `which.max` and `which.min`  functions reason why this parameter has no effect on these functions.

rdstern commented:
@MeSophie I agree, and think we do need a change in this sub-dialog:

![image](https://github.com/IDEMSInternational/R-Instat/assets/11226469/a21410df-843e-476b-bbbc-772df26c5339)

I suggest the controls are fine, but we are doing something rather different, when using max/min compared to the First, etc.   So I suggest we should not mix the 2 in the same summary.  So could you add 2 radio buttons with labels Display Positions and Use Positions.  
Then underneath the Display radio button there is a small group box with the max and min.  And the others are in a second group box. 

Then change the `Order by Another Variable` checkbox label to `Include Another Variable` And change the heading to just `Variable:`

Finally I agree that when we have the `where`, we don't also have the `which`.  So the max and min either give the overall row number, or (if the other variable is included) they give the variable at that row - so the variety with max yield, in the example above.

rdstern commented:
@MeSophie is this ready for me to check yet?

MeSophie commented:
> @MeSophie is this ready for me to check yet?

@rdstern  Sorry  i'm still on it.

MeSophie commented:
@lilyclements  Please I need your help with r codes on this PR.
Here is the current status of Columns Summaries > Summaries> Position

![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/57d46b11-c8ad-4312-ac40-833be21d440a)

And the which.max and which.min used in the summary_function.R in VB are:

```
summary_which_max <- function (x, na.rm = TRUE, na_type = "", ...) {
  if(length(x)==0 || (na.rm && length(x[!is.na(x)])==0)) return(NA)
  if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
  else{
    return(which.max(x))
  } 
}
summary_which_min <- function (x, na.rm = TRUE, na_type = "", ...) {
   if(length(x)==0 || (na.rm && length(x[!is.na(x)])==0)) return(NA)
  if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
  else{
    return(which.min(x))
  } 
}

```
Unfortunately these functions display the row number within the factor(s) and not the index of the overall rows.
Could  you please help me edit any of this code so that the result is the index of the overall rows?
In doing so please take into account the second variable which may or may not be included by the user much like the function `first`. 
Thank you in advance.

lilyclements commented:
@rdstern Can I just check what we want `summary_which_max` and `summary_which_min` to do? 
What would you want it to give as the output? Currently it gives the index, and I assume from this conversation you want to have the actual max/min value in the data, but perhaps I am mistaken?

Here's an example me using this function. It gives 15 as the minimum since the 15th value is our minimum (the value is 10.4). Can you give an example of what we would like to get here?

```
summary_which_min <- function (x, na.rm = TRUE, na_type = "", ...) {
  #TODO This prevents warning and Inf being retured. Is this desirable?
  if(length(x)==0 || (na.rm && length(x[!is.na(x)])==0)) return(NA)
  if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
  else return(which.min(x))
}

summary_which_min(mtcars$mpg)
```

rdstern commented:
Thanks @lilyclements I am hoping that which_min will give 15, as you say.  It is the row number of the minimum.  Then I would like where_min to give the 15th row of another variable, e.g. the make of the car that had the minimum.

MeSophie commented:
@rdstern, @lilyclements  Do we really need the `where max `function on this dialogue since it is give the same result as `max` function? 
![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/bd117b46-700f-4b59-82e3-7cacfd3652c1)


Also @lilyclements  the `which max` function you that you gave is the same one I used and it gives the results of the index within the factors and not the row indices in the overall data rows (what @rdstern  wants). 
For example, for our survey data we must have the column `New_whichMax` instead of `which_max` we actually have.
![image](https://github.com/IDEMSInternational/R-Instat/assets/106779233/b0759a5a-1fcd-4f7b-8413-021a309d03dd)

rdstern commented:
@MeSophie the where_max is only giving the same as max, because it is incorrect.  You are right that the which_max should give the overall row number.  Then the where_max should give the value of another specified variable at that row.  So, for your example the which>max (new) correctly gives row 15, and if you added the Variety variable, to the where_max, then you would know which variety we used in row 15.  I hope that's clear?

rdstern commented:
@MeSophie this has gone quiet, and I wonder what is happening?  I thought it was quite a simple step.  I think you also changed the formatting in the column summaries here?  With the delay others noticed the issue and then (I think) have also done that bit in #8943 .  Could you check that it is the same as you have done?  If so, then would you like them to take over this pull request so it is finished soon.  I also don't understand the json reference in the title here.

MeSophie commented:
> @MeSophie this has gone quiet, and I wonder what is happening? I thought it was quite a simple step. I think you also changed the formatting in the column summaries here? With the delay others noticed the issue and then (I think) have also done that bit in #8943 . Could you check that it is the same as you have done? If so, then would you like them to take over this pull request so it is finished soon. I also don't understand the json reference in the title here.

@rdstern sorry for the delay. The  json reference is because I added text on json file to solve translation issue #8767. I dont have any problem if they take over this PR.

rdstern commented:
Thanks @MeSophie .  You have quite a number of pending dialogs, so I would be happy for @derekagorhom to take this over.  Could you please detail what remains to be done to complete this pull request.

lilyclements commented:
@rdstern do we want to have the `summary_where_min` and `summary_where_max` in the sub-dialogs for the Column Summaries dialog?
In these summaries, we define two variables - `x` and `y`. I'm not sure how these summaries can work in a dialog like Column Summaries where we repeat the summary on each `x` variable (by `factor`), and so do not define a `y`.

rdstern commented:
@lilyclements I was hoping, just like with the calculator?  So the which_max, which I think is working, gives the row number of the maximum for each combination.
For the where max, you open the data frame as shown above and give the variable from there, (instead  of the row number(s) so it will give the value from that variable for the given row number(s).

Foe example if you are summarising the yields by village, then the which.max would tell you which row in each village had the maximum yield.  The where-max, with variety as the variable in the dialog above, would say which variety provided the maximum?

lilyclements commented:
> Foe example if you are summarising the yields by village, [...] the where-max, with variety as the variable in the dialog above, would say which variety provided the maximum?

@rdstern how do you see this being put on the dialog?
We have "yield" in the "Columns to Summarise" receiver
We have "village" in the "by" receiver
Where does "variety" go? Could it be a single receiver on the sub-dialog when you click on the "where-max" button?

rdstern commented:
@lilyclements that's exactly where it goes.

lilyclements commented:
@rdstern great - makes sense, I didn't realise that. @derekagorhom does this make sense to you to now work on?

lilyclements commented:
@MeSophie and @rdstern, Derrick noticed an issue in the `summary_which_min` and `summary_which_max` codes for when there are multiple minimums or maximums, it will return just one row. We've modified this code therefore to be the following: 

```
summary_which_max <- function (x, na.rm = TRUE, na_type = "", ...) {
  if(length(x)==0 || (na.rm && length(x[!is.na(x)])==0)) return(NA)
  if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
  else{
    # Get the minimum value
    max_value <- max(x, na.rm = na.rm)
    # Return all indices where x is equal to the minimum value
    return(which(x == max_value))
  } 
}

summary_which_min <- function(x, na.rm = TRUE, na_type = "", ...) {
  if(length(x) == 0 || (na.rm && length(x[!is.na(x)]) == 0)) return(NA)
  if(na.rm && na_type != "" && !na_check(x, na_type = na_type, ...)) return(NA)
  else {
    # Get the minimum value
    min_value <- min(x, na.rm = na.rm)
    # Return all indices where x is equal to the minimum value
    return(which(x == min_value))
  }
}
```

(e.g., if you now try for `summary_which_min(mtcars$mpg)` it will give both rows where there is a minimum, not just one, because the minimum runs over two rows). 

Is this okay?

rdstern commented:
@MeSophie and @derekagorhom and @N-thony I am no longer sure who is working on this, but please could it be done soon?  
I note that @lilyclements has, above, added the code for situations where which.max etc has multiple values.  I note that we already have the which.max etc in the summary keyboard of the calculator.  There, it just gets the row number of the first maximum, and I woul;d be happy if we do the same here.  So, let's not get too perfectionist, and just complete the simple task with the first maximum and the first minimum.

rdstern commented:
@MeSophie and @derekagorhom and @N-thony I am concerned that this needs to be completed - it started in February, and includes some tidying of the summary dialog.  I am not sure who is taking responsibility, but would like it to be completed this week if possible, while @lilyclements is around again.

lilyclements commented:
@MeSophie thank you for highlighting this problem. I've fixed the `summary_where_max` and `summary_where_min` functions to work with `$calculate_summary` (see PR [here](https://github.com/MeSophie/R-Instat/pull/244)).

Can you merge this in? Then to work on your side, the parameter name for the variable in the sub dialog is `summary_where_y`

```
data_book$calculate_summary(data_name="survey", columns_to_summarise=c("yield", "fert"), factors="village", j = 1,
                            summaries=c("summary_where_min", "summary_where_max", "summary_sum"),
                            summary_where_y = "variety",
                            silent=TRUE)
```


@rdstern what do we want to do if there are multiple minimum and maximums in `summary_which_min`, and `summary_which_max` - currently we only give the first instance since we give it as a single numerical value in the cell

MeSophie commented:
@rdstern could you please test the new changes? 
I had to add two extra check boxes for where max and where min because each control is linked to a single function. And I wonder if the` include another variable `check box is necessary for where max/min. Because it can cause confusion and lead the user to believe that this variable is optional, as is the case with the `First` function.
I think it would be a good to remove this check box and display the selector and the receiver as soon as where max or where min is ticked.
![image](https://github.com/user-attachments/assets/a069cf53-8551-4641-b329-ae601ccddd90)

rdstern reviewed (commented):
@MeSophie and @lilyclements that's looking good.  I tried with the survey data.
a) It was a surprise to me that the which_max was the row number within the factor level of the max, and not overall.  But I now think that may be ok.  I want to check that you were aware of this?  The reason it is now ok to me - if to you - is that if I want the row number overall, then I can get it by having a row variable (1 to 36) and using that in the where_max!
b) The one problem I found - logical is that with where_max or where_min you can do it without the other variable checked.  This given an error.  Could you check for this and then use the first variable in the data frame if none is specified in the dialog?
Then I'd really like to get this merged.  Perhaps @N-thony could check the code, while you are making this small change?

MeSophie commented:
@rdstern  the changes you are requesting require a modification to the R codes of the where and which functions. I have tried to b), the  only thing I've been able to do is display NA when the box is unchecked. @lilyclements Please once again I need your help on this PR.
![image](https://github.com/user-attachments/assets/89fd80f9-2057-47b4-8683-468b6b853213)

rdstern reviewed (approved):
@MeSophie that seems a nice fix.  I am approving.
@N-thony can you check and hopefully merge?

MeSophie commented on instat/sdgSummaries.vb:
```diff
@@ -562,27 +602,24 @@ Public Class sdgSummaries
+        ElseIf Not ucrChkWhereMax.Checked AndAlso Not ucrChkWhereMin.Checked Then
+            ucrChkInclude.Checked = False
+            ucrChkInclude.Enabled = False
+        End If
```
This is solved.

MeSophie commented on instat/sdgSummaries.vb:
```diff
@@ -645,4 +689,39 @@ Public Class sdgSummaries
+                clsDefaultFunction.RemoveParameterByName("summary_where_y")
+            End If
+        Else
+            clsDefaultFunction.RemoveParameterByName("-summary_where_y")
```
This is solved.

MeSophie commented on instat/sdgSummaries.vb:
```diff
@@ -592,15 +629,22 @@ Public Class sdgSummaries
+            End If
         Else
             clsDefaultFunction.RemoveParameterByName("order_by")
         End If
```
This is solved.

rdstern commented:
@lilyclements is going to do the final check on this.  It would then be great to get it merged.

lilyclements reviewed (changes requested):
@MeSophie this is great - nice one. Just one thought is that we currently have a checkbox "Include Another Variable" on the Positions dialog. 
In the case of "Where Max" and "Where Min" should the user always have to give another variable? It just gives NA otherwise, I presume always.
Is there a reason this is a checkbox, or can this checkbox instead be removed?

![image](https://github.com/user-attachments/assets/6a2d9f50-f0a0-49ed-889f-11444d3592fe)

MeSophie commented:
@lilyclements Roger's original intention was that if the user did not enter the other variable, then the summary_where_y argument would automatically take the variable(s) contained in the Summarise By: receiver. So it's a case of defining the factors variable as the default value for summary_where_y. I couldn't do it at the moment, the default value is NULL. I don't know how easy it will be for you to implement this.
Removing the check box is also a good idea as it would force the user to fill in the second variable as soon as Where Max or Where Min is checked.

lilyclements commented:
@MeSophie I understand. The user could always manually put that in for now? The `summarise_by` isn't read into this function really so that would be a larger change. What do you think?

MeSophie commented:
@lilyclements For now We can just delete the `Include Another Variable` box and make the selector  and receive appear if Where_Max or Where_Min is checked. I think that the user will then understand that he needs to fill in the Order By receiver. Because getting a result with the NA is not very useful.

MeSophie commented:
@lilyclements I removed the `Include Another Variable` check box when using where_max/where_min. Please could you check the dialog again? Thank you.

lilyclements reviewed (approved):
@MeSophie looks good. @N-thony over to you

rdstern commented:
@N-thony another important change in the Column Summaries dialog.  I hope you can check and merge.  I don't think it should interact, in any way with your drop unused levels change (#8986), but it may, as they are on the same dialog?

lilyclements commented:
@rdstern @N-thony if it helps, I merged these both into the same branch when testing (from laziness, not anything clever), and it didn't have conflicts!
